tratto/src/main/java/star/tratto/util/javaparser/JavaParserUtils.java

```diff
@@ -0,0 +1,1078 @@
+     *                      Must be false if checking {@code doesInstanceofCompile}. This is because
+     *                      type1 and type2 may be generics or unresolvable classes, and in those cases
+     *                      we cannot use the instanceof operator in a generated oracle, because it
+     *                      would not compile.
+     */
+    private static boolean isInstanceOf(String type1, String type2, OracleDatapoint oracleDatapoint, boolean checkEquality) {
+        // Preliminary checks
+        if (JavaTypes.PRIMITIVE_TYPES.contains(type1) || JavaTypes.PRIMITIVE_TYPES.contains(type2)) {
+            return false;
+        }
+        if (checkEquality && type1.equals(type2)) {
+            return true;
+        }
+        ResolvedType resolvedType2 = tryToGetResolvedType(type2);
+        if (resolvedType2 == null) {
+            return false; // Type2 is generic or an unknown class. In both cases, type1 cannot be instanceof it
+        }
+
+        String classSourceCode;
+        String className;
+        if (oracleDatapoint != null) {
+            classSourceCode = oracleDatapoint.getClassSourceCode();
+            className = oracleDatapoint.getClassName();
+        } else {
+            classSourceCode = SYNTHETIC_CLASS_SOURCE;
+            className = SYNTHETIC_CLASS_NAME;
+        }
+
+        // Parse class that contains method under test and add synthetic method to get resolved type 1 (we already have type 2)
+        CompilationUnit cu = javaParser.parse(classSourceCode).getResult().get();
+        BlockStmt syntheticMethodBody = getSyntheticMethod(cu, oracleDatapoint).getBody().get();
+        syntheticMethodBody.addStatement(type1 + " type1Var;");
+
+        // Get result of instanceof expression
+        try {
+            return resolvedType2.isAssignableBy(
+                    getClassOrInterface(cu, className)
+                            .getMethodsByName(SYNTHETIC_METHOD_NAME).get(0)
+                            .getBody().get()
+                            .getStatements().getFirst().get()
+                            .asExpressionStmt().getExpression()
+                            .asVariableDeclarationExpr().getVariables().get(0)
+                            .resolve().getType());
+        } catch (UnsolvedSymbolException e) {
+            logger.warn("Failed to evaluate instanceof within method. Expression: \"type1Var instanceof {}\". Method: {}", type2, syntheticMethodBody.getParentNode().get());
+            return false;
+        }
+    }
+
+    private static ResolvedType tryToGetResolvedType(String type2) {
+        ResolvedType resolvedType2;
+        try {
+            resolvedType2 = getResolvedType(type2);
+        } catch (UnsolvedSymbolException e) {
+            return null;
+        }
+        return resolvedType2;
+    }
+
+    // TODO: The "not the other way around" is true only in a certain sense.  Given variables `b`
```

Comment:
Please address this todo comment.  What is the spec of this method?  Does it allow boxing but not unboxing?  Why is such a specification important to clients?

Reply:
@AML14